src/lib/reasoners/satml_frontend.ml

```diff
@@ -1095,6 +1202,27 @@ module Make (Th : Theory.S) : Sat_solver_sig.S = struct
           | Util.Timeout -> model_gen_on_timeout env
         end
 
+  let rec unsat_rec_prem env ~first_call : unit =
+    try
+      unsat_rec env ~first_call
+    with
+    | I_dont_know env as e ->
+      begin
+        try analyze_unknown_for_objectives env e unsat_rec_prem
+        with
+        | IUnsat (env, _) ->
+          assert (!(env.objectives) != None);
+          (* objectives is a ref, it's necessiraly updated as a
+             side-effect to best value *)
+          raise (I_dont_know env)
+      end
+    | Util.Timeout as e -> raise e
+
+    | IUnsat (env, _) as e ->
+      if !(env.objectives) == None then raise e;
+      (* TODO: put the correct objectives *)
```

Comment:
Yeah, that matches my understanding as well. But I think that this is actually unrelated to `env.objectives` —  I possibly this is related to the fact that the env is not updated? As in, the `env.objectives` is correct (in the sense that it is the optimum) but the `env` that we return in `I_dont_know` does *not* satisfy the objectives anymore (since it is an unsat environment). And model generation still works because when we compute the model we also use a reference to a lazy term that uses the *last known sat env* instead of using the current env…

Reply:
Good remark. I will investigate on it later. I think we should merge now :)